### ISSUE TITLE
Incorrect proofs with 'old' keyword?

### ISSUE THREAD

**Issue by rec0de** [Edkamb/crowbar-tool#15]
I may have found a problem with how the 'old' keyword is implemented. I think it's best described by the following example:

```module OldBug;

data Spec = Ensures(Bool);

class Testcase {
  Int dontTouch = 0;

  [Spec: Ensures(this.dontTouch == old(this.dontTouch))]
  Unit modify() {
    this.dontTouch = 42;
    this.doNothing();
  }

  [Spec: Ensures(this.dontTouch == old(this.dontTouch))]
  Unit doNothing() { }
}

{}
```
This program passes verification in the current version of crowbar. The specification of method `doNothing` is obviously correct and verifies as expected. Method `modify`, however, obviously changes the value of `dontTouch` - but it seems like the postcondition from method `doNothing` is used as a premiss in the verification of `modify` without accounting for the fact that the `old` in one method refers to a different state that the `old` in the other.

Adding a `Requires(this.dontTouch == old(this.dontTouch))` to the `doNothing` method fixes the problem.

Perhaps my interpretation of what `old` is designed to do is wrong, but I would argue that the behaviour shown in the example is very counter-intuitive.

**Comment by Edkamb** [comment]
This is indeed a bug, old must be translated differently when used in a post-condition, I will look at this next.

**Comment by rec0de** [comment]
I think the fix in e1f2f5e has an issue in the update order: *old* in the postcondition of the called method points to the post-anonymization heap because the `anon` update is applied after the `{old := heap}` update. This means we can't verify the above testcase even if the modification to the `dontTouch` field is removed because the postcondition of `doNothing` essentially reads `select(anon(heap), dontTouch) = select(anon(heap), dontTouch)` after the updates are applied.

Suggested change to `PITSyncCallAssign` (l.327-337):

```kotlin
val postCond = repos.syncMethodEnss[call.met]?.first ?: True
val targetPostDecl = repos.syncMethodEnss[call.met]!!.second
val substPostMap = mapSubstPar(call, targetPostDecl)

val anon = ElementaryUpdate(Heap, anon(Heap))

 //XXX handle  last here as well
 val someHeap = FreshGenerator.getFreshProgVar(Heap.dType)
 val heapUpdate = ChainUpdate(ElementaryUpdate(OldHeap,Heap), ElementaryUpdate(LastHeap,someHeap))

 val updateLeftNext = ChainUpdate(input.update, ChainUpdate(heapUpdate, ChainUpdate(anon, updateNew)))
```
